rfcs/2021-01-08-5843-encdoing-decoding-format-vector.md

```diff
@@ -0,0 +1,95 @@
+
+## Rationale
+
+Unfortunately in most of my research I've pretty well come to the conclusion that encoding formats as a domain are rife with technical and political failures. That is to say, in some regard or another, every format from JSON to Flatbuffers is going to force a tradeoff between correctness, performance, features or tooling and maintenance overhead. Inlined below here is a reasonable starting point on performance. It's taken from the [NoProto project](https://https://github.com/only-cliches/NoProto) since they're regularly updating their benchmarks. They've also done an excellent job of enumerating some of the aspects to consider when making a decision of this nature and their read on the space very much mirrors my own thinking in many ways.
+
+For the below data, Encodes and Decodes are ops/sec so higher is better while size before and after compression lower is better. This list is far from exhaustive and specifcially lacks a serde_json comparison which is unfortunate but other benches out there do include them (though obviously a different benchmark makes cross comparison difficult). 
+
+| Library            | Encode | Decode All | Decode 1 | Update 1 | Size (bytes) | Size (Zlib) |
+|--------------------|--------|------------|----------|----------|--------------|-------------|
+| **Runtime Libs**   |        |            |          |          |              |             |
+| *NoProto*          |   1057 |       1437 |    47619 |    12195 |          208 |         166 |
+| Apache Avro        |    138 |         51 |       52 |       37 |          702 |         336 |
+| FlexBuffers        |    401 |        855 |    23256 |      264 |          490 |         309 |
+| JSON               |    550 |        438 |      544 |      396 |          439 |         184 |
+| BSON               |    115 |        103 |      109 |       80 |          414 |         216 |
+| MessagePack        |    135 |        222 |      237 |      119 |          296 |         187 |
+| **Compiled Libs**  |        |            |          |          |              |             |
+| Flatbuffers        |   1046 |      14706 |   250000 |     1065 |          264 |         181 |
+| Bincode            |   5882 |       8772 |     9524 |     4016 |          163 |         129 |
+| Protobuf           |    859 |       1140 |     1163 |      480 |          154 |         141 |
+| Prost              |   1225 |       1866 |     1984 |      962 |          154 |         142 |
+
+Performance is a single aspect of a nuanced problem for us; for more perf benchmarks I'd suggest checking out [rust-serialization-benchmarks](https://github.com/erickt/rust-serialization-benchmarks).  Some of these formats (JSON for example) still leaves us with a marginal concern around both dropped metadata as well as incidental exposure of a private data format to our users. The goal here is to hopefully solve for all of these things with whatever path is chosen.
```

Comment:
I'm not really worried about exposing the data format to users and, as an open source project, we do that by default. We can document that it is internal and may change at any time.